src/stores/A_store.js

```diff
@@ -60,15 +63,89 @@ class A_store extends ReduceStore {
+                        player: {x: -1, y: -1},
+                        end: {x: -2, y: -2}, 
+                        stage: 'STARTP', 
+                        path: null, 
+                        visibleCells: null, 
+                        seenCells: null
+                    };
             }
 
             case ActionTypes.STEP: {
-                const path = [...state.path];
-                const nextLocation = path.shift();
+                const [nextLocation, ...path] = state.path;
 
                 const stage = path.length > 0 ? 'STEP' : 'RESET';
-                return {...state, stage, path, player: {x: nextLocation.x, y: nextLocation.y}};
+                return {
+                    ...state, 
+                        stage, 
+                        path, 
+                        player: {x: nextLocation.x, y: nextLocation.y}
+                    };
+            }
+
+            case ActionTypes.GENERATE_FOG: {
+                const {player, end} = state;
+                const {x, y} = player;
+                const revealedGrid = mapGenerator(player, end, new Grid(SIZE, SIZE));
+                const pathGrid = new Grid(SIZE, SIZE);
+
+                const visibleCells = pathGrid.getVisible(x, y);
+                revealedGrid.copyCells(visibleCells);
+                const seenCells = pathGrid.getVisible(x, y);
+
+                const start = pathGrid.getCell(x, y);
+                const goal = pathGrid.getCell(end.x, end.y);
+                const path = Astar(start, goal, Cell.heuristic);
+                path.shift();
+
+                return {
+                        ...state, 
+                        pathGrid, 
+                        revealedGrid,
+                        path, 
+                        stage: 'STEP_FOG', 
+                        visibleCells, 
+                        seenCells
+                    };
+            }
+
+            case ActionTypes.STEP_FOG: {
+                let [nextLocation, ...path] = state.path;
+                const {revealedGrid, end, seenCells} = state;
+                let {pathGrid} = state;
+                let stage;
+                let visibleCells;
+
+                if (path.length > 0) {
+                    stage = 'STEP_FOG';
+                    visibleCells = pathGrid.getVisible(nextLocation.x, nextLocation.y);
+                    revealedGrid.copyCells(visibleCells);
+
+                    for (const cell of visibleCells) {
+                        seenCells.add(cell);
+                    }
+
+                    const start = pathGrid.getCell(nextLocation.x, nextLocation.y);
+                    const goal = pathGrid.getCell(end.x, end.y);
+                    path = Astar(start, goal, Cell.heuristic);
+                    path.shift();
+                } else {
+                    stage = 'RESET';
+                    pathGrid = revealedGrid;
+                }
+
+                return {
+                        ...state, 
```

Comment:
at least be consistent within your own changeset <img src="https://cdn.discordapp.com/emojis/390637045678342144.png" alt=":blobcatverysad:" width="22" height="22">

*one level of indent* for the properties, ending brace is on same level as the line with starting brace

again, just look at `getInitialState`